topology/probes/nsm/nsm.go

```diff
@@ -0,0 +1,395 @@
+				dst:      conn.GetLocalDestination(),
+			},
+		}
+		logging.GetLogger().Infof("adding link %+v", l)
+		p.connections = append(p.connections, l)
+		p.g.Lock()
+		l.AddEdge(p.g)
+		p.g.Unlock()
+	} else {
+		for i := range p.connections {
+			l, ok := p.connections[i].(*localConnectionPair)
+			if !ok {
+				continue
+			}
+			if l.srcInode == srcInode && l.dstInode == dstInode {
+				p.g.Lock()
+				l.DelEdge(p.g)
+				p.g.Unlock()
+				p.removeConnectionItem(i)
+				break
+			}
+		}
+	}
+}
+
+func (p *Probe) onConnLocalRemote(t cc.CrossConnectEventType, conn *cc.CrossConnect) {
+	p.Lock()
+	defer p.Unlock()
+	c, i := p.getConnectionWithRemote(conn.GetRemoteDestination().GetId())
+	if t != cc.CrossConnectEventType_DELETE {
+		if c == nil {
+			c = &remoteConnectionPair{
+				baseConnectionPair: baseConnectionPair{
+					src:     conn.GetLocalSource(),
+					payload: conn.GetPayload(),
+				},
+				remote: conn.GetRemoteDestination(),
+				srcID:  conn.GetId(),
+				//TODO: compare payloads, they should be identical
+			}
+			p.connections = append(p.connections, c)
+			//TODO: we erase previous payload if exist
+			c.payload = conn.GetPayload()
+
+		} else {
+			c.src = conn.GetLocalSource()
+			c.srcID = conn.GetId()
+			//TODO: we erase previous payload if exist
+			c.payload = conn.GetPayload()
+			p.g.Lock()
+			c.AddEdge(p.g)
+			p.g.Unlock()
+		}
+	} else {
+		if c == nil {
+			logging.GetLogger().Warning("NSM: received cross connect delete event for a connection that does not exist")
+			return
+		}
+
+		p.g.Lock()
+		c.DelEdge(p.g)
+		p.g.Unlock()
+
+		// Delete the link only if dst is also empty
+		if c.dst == nil {
+			p.removeConnectionItem(i)
+		} else {
+			c.src = nil
+		}
+	}
+}
+
+func (p *Probe) onConnRemoteLocal(t cc.CrossConnectEventType, conn *cc.CrossConnect) {
+	p.Lock()
+	defer p.Unlock()
+	c, i := p.getConnectionWithRemote(conn.GetRemoteSource().GetId())
+	if t != cc.CrossConnectEventType_DELETE {
+		if c == nil {
+			c = &remoteConnectionPair{
+				baseConnectionPair: baseConnectionPair{
+					dst:     conn.GetLocalDestination(),
+					payload: conn.GetPayload(),
+				},
+				remote: conn.GetRemoteSource(),
+				dstID:  conn.GetId(),
+				//TODO: compare payloads, they should be identical
+			}
+			p.connections = append(p.connections, c)
+		} else {
+			c.dst = conn.GetLocalDestination()
+			c.dstID = conn.GetId()
+			//TODO: compare payloads, they should be identical
+			c.payload = conn.GetPayload()
+			p.g.Lock()
+			c.AddEdge(p.g)
+			p.g.Unlock()
+		}
+	} else {
+		if c == nil {
+			logging.GetLogger().Warning("NSM: received cross connect delete event for a connection that does not exist")
+			return
+
+		}
+
+		p.g.Lock()
+		c.DelEdge(p.g)
+		p.g.Unlock()
+
+		// Delete the link only if src is also empty
+		if c.GetSource() == nil {
+			p.removeConnectionItem(i)
+		} else {
+			c.dst = nil
+		}
+	}
+}
+
+// OnNodeAdded event
+// We assume skydive has locked the graph before calling this function
```

Comment:
nit: correct, therefor this comment can safely be removed :-)

Reply:
fine, that was just a reminder :)